### PR TITLE
Fix #23567

### LINKED ISSUE #23567: Are isbits-Unions isbits?
Noticed in the current master bug:
```julia
julia> t = Vector{Union{Float64, Void}}(5)
5-element Array{Union{Void, Float64},1}:
 nothing
 nothing
 nothing
 nothing
 nothing

julia> b = collect(Union{Float64, Void}, 1.0:3.0)
3-element Array{Union{Void, Float64},1}:
 1.0
 2.0
 3.0

julia> copy!(t, 2, b)
5-element Array{Union{Void, Float64},1}:
 0.0
 1.0
 2.0
  nothing
  nothing
```
The issue is that we currently go thru the codepath in array.jl
```julia
function unsafe_copy!(dest::Array{T}, doffs, src::Array{T}, soffs, n) where T
    if isbits(T)
        unsafe_copy!(pointer(dest, doffs), pointer(src, soffs), n)
    else
        ccall(:jl_array_ptr_copy, Void, (Any, Ptr{Void}, Any, Ptr{Void}, Int),
              dest, pointer(dest, doffs), src, pointer(src, soffs), n)
    end
```
but `isbits(Union{Float64, Void})` is false; so we hit `jl_array_ptr_copy`. Now, I did had a hook in `jl_array_ptr_copy` to intercept isbits-Union arrays and handle them specially, the problem is it currently only works when you happen to be copying to the 1st element of the dest array. This is because we only pass `dest_p`, a pointer and not an index, to `jl_array_ptr_copy`, so we don't actually know where to copy selector bytes from src to dest.

There a couple of ways to address this:
- go ahead and make `isbits(Union{Float64, Void}) == true`, this would involve making sure we're careful in a # of places in Base where we're currently checking `isbits(T)` on arrays and doing optimizations (i.e. `vcat`, `unsafe_copy`, `flipdim`). The biggest ? for me in doing this is how this propagates thru the rest of the system, in particular, inference.jl (which has quite a few calls to `isbits(T)`.). Overall, I think it would be more correct, but it does seem a bit worrisome not quite knowing the full extent of this kind of change. I can certainly do more research here if we think it's the best way to try out first.
- make another method `jl_array_ptr_copy2`, which also takes `ind` arguments so we can correctly adjust when copying isbits-Union arrays.

Ultimately, I think we _should_ make isbits-Unions isbits, because we're essentially doing that everywhere in `src/`, so we might as well be consistent in Base.

cc: @JeffBezanson, @vtjnash

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -181,6 +181,14 @@ the same manner as C.
...
+    elseif isbitsunion(T)
+        ccall(:memmove, Ptr{Void}, (Ptr{Void}, Ptr{Void}, UInt),
+              pointer(dest, doffs), pointer(src, soffs), n * Base.bitsunionsize(T))
+        # copy selector bytes
+        ccall(:memmove, Ptr{Void}, (Ptr{Void}, Ptr{Void}, UInt),
+              convert(Ptr{UInt8}, pointer(dest)) + length(dest) * Base.bitsunionsize(T) + doffs - 1,
+              convert(Ptr{UInt8}, pointer(src)) + length(src) * Base.bitsunionsize(T) + soffs - 1,
+              n)
     else
         ccall(:jl_array_ptr_copy, Void, (Any, Ptr{Void}, Any, Ptr{Void}, Int),
```
Comment (yuyichao) [inline base/array.jl]: Remove the branch from the C function?

Comment (quinnj) [reply base/array.jl]: Oh, thanks for reminding me; just pushed that.

**Comment by vtjnash** [comment]
This looks mostly ok, at least as an improvement over the current issues. Note however, that `isbitsunion` is a very, very slow and inefficient query – and also is effectively just as incorrect as branching on `is bits`. But to avoid splitting the discussion, comments and design should probably happen on #23567 instead of here.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/23577#issuecomment-327031080):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/686256969e542f1c97a1e90862cba089b3c9b664/628a81d_vs_fdef237/report.md)
cc @ararslan

**Comment by quinnj** [comment]
I'm guessing the curl compile error is unrelated 😄 . But on a related node, does nanosoldier build julia from scratch for each benchmark run? That seems intense! (if not excessive!)

**Comment by ararslan** [comment]
Yeah, Nanosoldier builds from scratch on every run, which is why the runs can take so long.

**Pull request merged**